overlord/udevmon.go

```diff
@@ -0,0 +1,141 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package overlord
+
+import (
+	"fmt"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/osutil/udev/netlink"
+
+	"github.com/snapcore/snapd/interfaces/hotplug"
+	"github.com/snapcore/snapd/logger"
+)
+
+type UDevMon interface {
+	Connect() error
+	Run() error
+	Stop() error
+}
+
+type DeviceAddedCallback func(device *hotplug.HotplugDeviceInfo)
+type DeviceRemovedCallback func(device *hotplug.HotplugDeviceInfo)
+
+// UDevMonitor monitors kernel uevents making it possible to find USB devices.
+type UDevMonitor struct {
+	tmb             *tomb.Tomb
+	deviceAddedCb   DeviceAddedCallback
+	deviceRemovedCb DeviceRemovedCallback
+	netlinkConn     *netlink.UEventConn
+	// channels used by netlink connection and monitor
+	monitorStop   chan struct{}
+	netlinkErrors chan error
+	netlinkEvents chan netlink.UEvent
+}
+
+func NewUDevMonitor(addedCb DeviceAddedCallback, removedCb DeviceRemovedCallback) UDevMon {
+	m := &UDevMonitor{
+		deviceAddedCb:   addedCb,
+		deviceRemovedCb: removedCb,
+		netlinkConn:     &netlink.UEventConn{},
+		tmb:             new(tomb.Tomb),
+	}
+
+	m.netlinkEvents = make(chan netlink.UEvent)
+	m.netlinkErrors = make(chan error)
+
+	return m
+}
+
+func (m *UDevMonitor) Connect() error {
+	if m.netlinkConn == nil || m.netlinkConn.Fd != 0 {
+		// this cannot happen in real code but may happen in tests
+		panic("cannot run UDevMonitor more than once")
```

Comment:
I wonder if we shouldn't simply return an error here like "cannot connect: already connected" or something?